File: veinminer-enchant/src/main/java/de/miraculixx/veinminerEnchant/paper/VeinminerEnchantmentSettings.java
package de.miraculixx.veinminerEnchant.paper;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;

import java.nio.file.Files;
import java.nio.file.Path;

public record VeinminerEnchantmentSettings(int minCost, int maxCost, int anvilCost, boolean pickaxeOnly) {
    private static final Path FILE_PATH = Path.of("plugins", "Veinminer", "enchantmentSettings.json");
    private static final Gson GSON = new GsonBuilder().setPrettyPrinting().create();

    public VeinminerEnchantmentSettings() {
        this(15, 65, 7, false);
    }

    public static VeinminerEnchantmentSettings get() {
        VeinminerEnchantmentSettings defaults = new VeinminerEnchantmentSettings();

        try {
            if (!Files.exists(FILE_PATH)) {
                Path parent = FILE_PATH.getParent();
                if (parent != null) Files.createDirectories(parent);
                Files.writeString(FILE_PATH, GSON.toJson(defaults));
                System.out.println("[VeinminerEnchant] Created " + FILE_PATH.getFileName() + " default config");
                return defaults;
            }

            RawSettings raw = GSON.fromJson(Files.readString(FILE_PATH), RawSettings.class);
            return new VeinminerEnchantmentSettings(
                    raw != null && raw.minCost != null ? raw.minCost : defaults.minCost,
                    raw != null && raw.maxCost != null ? raw.maxCost : defaults.maxCost,
                    raw != null && raw.anvilCost != null ? raw.anvilCost : defaults.anvilCost,
                    raw != null && raw.pickaxeOnly != null ? raw.pickaxeOnly : defaults.pickaxeOnly
            );
        } catch (Exception e) {
            System.out.println("[VeinminerEnchant] Failed to load " + FILE_PATH.getFileName() + " config: Reason: " + e.getMessage());
            return defaults;
        }
    }

    private static final class RawSettings {
        Integer minCost;
        Integer maxCost;
        Integer anvilCost;
        Boolean pickaxeOnly;
    }
}

File: veinminer-enchant/src/main/kotlin/de/miraculixx/veinminerEnchant/paper/VeinminerEnchantmentSettings.kt
package de.miraculixx.veinminerEnchant.paper

import com.google.gson.GsonBuilder
import java.nio.file.Path
import kotlin.io.path.Path
import kotlin.io.path.createDirectories
import kotlin.io.path.exists
import kotlin.io.path.readText
import kotlin.io.path.writeText

data class VeinminerEnchantmentSettings(
    val minCost: Int = 15,
    val maxCost: Int = 65,
    val anvilCost: Int = 7,
    val pickaxeOnly: Boolean = false
) {
    companion object {
        private val filePath: Path = Path("plugins/Veinminer/enchantmentSettings.json")
        private val gson = GsonBuilder().setPrettyPrinting().create()

        private data class RawSettings(
            val minCost: Int? = null,
            val maxCost: Int? = null,
            val anvilCost: Int? = null,
            val pickaxeOnly: Boolean? = null
        )

        fun get(): VeinminerEnchantmentSettings {
            val defaults = VeinminerEnchantmentSettings()

            return try {
                if (!filePath.exists()) {
                    filePath.parent?.createDirectories()
                    filePath.writeText(gson.toJson(defaults))
                    println("[VeinminerEnchant] Created ${filePath.fileName} default config")
                    defaults
                } else {
                    val raw = gson.fromJson(filePath.readText(), RawSettings::class.java)
                    VeinminerEnchantmentSettings(
                        minCost = raw?.minCost ?: defaults.minCost,
                        maxCost = raw?.maxCost ?: defaults.maxCost,
                        anvilCost = raw?.anvilCost ?: defaults.anvilCost,
                        pickaxeOnly = raw?.pickaxeOnly ?: defaults.pickaxeOnly
                    )
                }
            } catch (e: Exception) {
                println("[VeinminerEnchant] Failed to load ${filePath.fileName} config: Reason: ${e.message}")
                defaults
            }
        }
    }
}
